### PR TITLE
Improvement to MPISpace and SHMEMSpace Operations 

### PR DESCRIPTION
**Pull request by janciesko**
- Adds atomic View support for MPISpace
- Removes use of `MPI_Type_size` in favor of `sizeof`
- Re-enables unit testing for atomic Views and MPISpace
- Minor fixes

### CODE REVIEW

File: src/impl/mpispace/Kokkos_MPISpace_Ops.hpp [hunk 1]
```diff
@@ -72,106 +68,135 @@ KOKKOS_REMOTESPACES_G(unsigned short, MPI_UNSIGNED_SHORT)
...
-KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned int)
-KOKKOS_REMOTESPACES_ATOMIC_SET(long)
-KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned long)
-KOKKOS_REMOTESPACES_ATOMIC_SET(long long)
-KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned long long)
-KOKKOS_REMOTESPACES_ATOMIC_SET(float)
-KOKKOS_REMOTESPACES_ATOMIC_SET(double)
+KOKKOS_REMOTESPACES_ATOMIC_SET(int, MPI_INT)
+KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned int, MPI_UNSIGNED)
+KOKKOS_REMOTESPACES_ATOMIC_SET(long, MPI_INT64_T)
```
Comment (devreal) [inline src/impl/mpispace/Kokkos_MPISpace_Ops.hpp:89]:
```suggestion
KOKKOS_REMOTESPACES_ATOMIC_SET(long, MPI_LONG)
```

File: src/impl/mpispace/Kokkos_MPISpace_Ops.hpp [hunk 2]
```diff
@@ -44,25 +42,23 @@ KOKKOS_REMOTESPACES_P(unsigned short, MPI_UNSIGNED_SHORT)
 KOKKOS_REMOTESPACES_P(int, MPI_INT)
 KOKKOS_REMOTESPACES_P(unsigned int, MPI_UNSIGNED)
 KOKKOS_REMOTESPACES_P(long, MPI_INT64_T)
```
Comment (devreal) [inline src/impl/mpispace/Kokkos_MPISpace_Ops.hpp:44]:
```suggestion
KOKKOS_REMOTESPACES_P(long, MPI_LONG)
```

File: src/impl/mpispace/Kokkos_MPISpace_Ops.hpp [hunk 3]
```diff
@@ -72,106 +68,135 @@ KOKKOS_REMOTESPACES_G(unsigned short, MPI_UNSIGNED_SHORT)
...
+    MPI_Fetch_and_op(&val, &ret, mpi_type, pe,                               \
+                     sizeof(SharedAllocationHeader) + offset * sizeof(type), \
+                     MPI_OP_NULL, win);                                      \
+    MPI_Win_flush(pe, win);                                                  \
+    return ret;                                                              \
+  }
+
+KOKKOS_REMOTESPACES_ATOMIC_FETCH(int, MPI_INT)
+KOKKOS_REMOTESPACES_ATOMIC_FETCH(unsigned int, MPI_UNSIGNED)
+KOKKOS_REMOTESPACES_ATOMIC_FETCH(long, MPI_INT64_T)
```
Comment (devreal) [inline src/impl/mpispace/Kokkos_MPISpace_Ops.hpp:111]:
```suggestion
KOKKOS_REMOTESPACES_ATOMIC_FETCH(long, MPI_LONG)
```

File: src/impl/mpispace/Kokkos_MPISpace_Ops.hpp [hunk 4]
```diff
@@ -72,106 +68,135 @@ KOKKOS_REMOTESPACES_G(unsigned short, MPI_UNSIGNED_SHORT)
...
+    MPI_Win_flush(pe, win);                                                   \
+  }
+
+KOKKOS_REMOTESPACES_ATOMIC_ADD(char, MPI_SIGNED_CHAR)
+KOKKOS_REMOTESPACES_ATOMIC_ADD(unsigned char, MPI_UNSIGNED_CHAR)
+KOKKOS_REMOTESPACES_ATOMIC_ADD(short, MPI_SHORT)
+KOKKOS_REMOTESPACES_ATOMIC_ADD(unsigned short, MPI_UNSIGNED_SHORT)
+KOKKOS_REMOTESPACES_ATOMIC_ADD(int, MPI_INT)
+KOKKOS_REMOTESPACES_ATOMIC_ADD(unsigned int, MPI_UNSIGNED)
+KOKKOS_REMOTESPACES_ATOMIC_ADD(long, MPI_INT64_T)
```
Comment (devreal) [inline src/impl/mpispace/Kokkos_MPISpace_Ops.hpp:135]:
```suggestion
KOKKOS_REMOTESPACES_ATOMIC_ADD(long, MPI_LONG)
```

File: src/impl/mpispace/Kokkos_MPISpace_Ops.hpp [hunk 5]
```diff
@@ -72,106 +68,135 @@ KOKKOS_REMOTESPACES_G(unsigned short, MPI_UNSIGNED_SHORT)
...
 
-KOKKOS_REMOTESPACES_ATOMIC_SET(int)
-KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned int)
-KOKKOS_REMOTESPACES_ATOMIC_SET(long)
-KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned long)
-KOKKOS_REMOTESPACES_ATOMIC_SET(long long)
-KOKKOS_REMOTESPACES_ATOMIC_SET(unsigned long long)
-KOKKOS_REMOTESPACES_ATOMIC_SET(float)
-KOKKOS_REMOTESPACES_ATOMIC_SET(double)
+KOKKOS_REMOTESPACES_ATOMIC_SET(int, MPI_INT)
```
Comment (devreal) [inline src/impl/mpispace/Kokkos_MPISpace_Ops.hpp:87]: Missing char and short versions

File: src/impl/mpispace/Kokkos_MPISpace_Ops.hpp [hunk 6]
```diff
@@ -72,106 +68,135 @@ KOKKOS_REMOTESPACES_G(unsigned short, MPI_UNSIGNED_SHORT)
...
 
-KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(int)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(unsigned int)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(long)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(unsigned long)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(long long)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(unsigned long long)
+KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(int, MPI_INT)
+KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(unsigned int, MPI_UNSIGNED)
+KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(long, MPI_INT64_T)
```
Comment (devreal) [inline src/impl/mpispace/Kokkos_MPISpace_Ops.hpp:157]:
```suggestion
KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(long, MPI_LONG)
```

File: src/impl/mpispace/Kokkos_MPISpace_Ops.hpp [hunk 7]
```diff
@@ -72,106 +68,135 @@ KOKKOS_REMOTESPACES_G(unsigned short, MPI_UNSIGNED_SHORT)
...
+    return ret;                                                              \
   }
 
-KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(int)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(unsigned int)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(long)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(unsigned long)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(long long)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(unsigned long long)
+KOKKOS_REMOTESPACES_ATOMIC_FETCH_ADD(int, MPI_INT)
```
Comment (devreal) [inline src/impl/mpispace/Kokkos_MPISpace_Ops.hpp:155]: Missing char and short

File: src/impl/mpispace/Kokkos_MPISpace_Ops.hpp (hunk 6 above)
Comment (devreal) [reply src/impl/mpispace/Kokkos_MPISpace_Ops.hpp:157]: It's everywhere...

**Review by devreal — CHANGES_REQUESTED** [review]
Two issues:

- Wrong use of `MPI_INT64_T` for `long`
- Missing `char` and `short` version in some places

File: src/impl/mpispace/Kokkos_MPISpace_Ops.hpp [hunk 8]
```diff
@@ -71,108 +66,155 @@ KOKKOS_REMOTESPACES_G(short, MPI_SHORT)
...
-KOKKOS_REMOTESPACES_ATOMIC_FETCH(float)
-KOKKOS_REMOTESPACES_ATOMIC_FETCH(double)
-
+#define KOKKOS_REMOTESPACES_ATOMIC_FETCH(type, mpi_type)                     \
+  static KOKKOS_INLINE_FUNCTION type mpi_type_atomic_fetch(                  \
+      type val, int offset, int pe, const MPI_Win &win) {                    \
+    type ret;                                                                \
+    MPI_Fetch_and_op(&val, &ret, mpi_type, pe,                               \
+                     sizeof(SharedAllocationHeader) + offset * sizeof(type), \
+                     MPI_OP_NULL, win);                                      \
```
Comment (devreal) [inline src/impl/mpispace/Kokkos_MPISpace_Ops.hpp:106]: This should be `MPI_NO_OP`, not `MPI_OP_NULL`. Besides that, why do you use an atomic operation for `operator-` etc?

Comment (janciesko) [reply src/impl/mpispace/Kokkos_MPISpace_Ops.hpp:106]: Correcting to MPI_NO_OP. We implemented those operators with an atomic fetch in order to preserve correct ordering (acquire memory semantics) with other operations.

**Review by devreal — APPROVED** [review]

**Pull request merged**